### PR TITLE
Keeper and redemption

### PR DESCRIPTION
**Pull request by ianballard**

### CODE REVIEW

File: contracts/TokenMaturity.sol [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        TokenMaturityLib.StateStorage storage ss = TokenMaturityLib.getState();
+        return ss.isInit;
+    }
+
+    function setContractMaturityTimestamp(uint256 timestamp) public {
+        TokenMaturityLib.StateStorage storage ss = TokenMaturityLib.getState();
+        ss.contractMaturityTimestamp = timestamp;
+    }
+
+    function isTokenMatured(uint256 timestamp) public view returns (bool) {
```
[the remark this answers is not in the document]
Comment (ianballard) [reply contracts/TokenMaturity.sol:41]: yeah i made it this way for easier testing. I'm fine with changing it if you are

File: contracts/TokenMaturity.sol [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        require(!hasRedeemed, "Address can only redeem once.");
+
+        TuffToken tuffToken = TuffToken(payable(address(this)));
+        uint256 ownerBalance = tuffToken.balanceOf(account);
+
+        require(ownerBalance > 0, "Owner balance needs to be greater than 0.");
+
+        uint256 redemptionAmount = getRedemptionAmount(account, ownerBalance);
+
+        account.transfer(redemptionAmount);
```
[the remark this answers is not in the document]
Comment (ianballard) [reply contracts/TokenMaturity.sol:102]: absolutely. I was planning on doing a full sweep of the dapp and add in events in many places. I'll do that in a separate ticket if thats okay with you.

File: test/tokenMaturity.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+const {expect} = require("chai");
+const hre = require("hardhat");
+
+const consts = require("../consts");
+const {BN} = require("@openzeppelin/test-helpers");
+const {randomBytes} = require('crypto');
+const utils = require("./utils");
+
+describe('TokenMaturity', function () {
+    this.timeout(5000);
```
[the remark this answers is not in the document]
Comment (ianballard) [reply test/tokenMaturity.js:12]: none this was copy pasta. will remove

File: contracts/v7/MarketTrend.sol [hunk 4]
```diff
@@ -95,54 +118,56 @@ contract MarketTrend {
...
-            !ss.trackingPeriods[ss.trackingPeriods.length - 1].isActive,
-            "Can only have one active tracking period."
-        );
+        delete ss.priceDataEntries;
 
         MarketTrendLib.PriceData memory priceData;
         priceData.price = getPrice();
-        priceData.timestamp = baseTrackingPeriodStart;
+        priceData.timestamp = _trackingPeriodStart;
         ss.priceDataEntries.push(priceData);
```
[the remark this answers is not in the document]
Comment (ianballard) [reply contracts/v7/MarketTrend.sol:121]: the delete keyword deletes all elements in an array not the array instance itself

File: contracts/TokenMaturity.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (unibeck) [reply contracts/TokenMaturity.sol:41]: I am all for easier testing. The pros outweighs the cons so let's keep it as is

File: contracts/TokenMaturity.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (unibeck) [reply contracts/TokenMaturity.sol:102]: Yeah, let's leave it for a future ticket. Can you add a `TODO` before merge?

**Review by unibeck — APPROVED** [review]

File: contracts/TokenMaturity.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (ianballard) [reply contracts/TokenMaturity.sol:102]: done

**Pull request merged**